src/createEmbeddingText.ts

```diff
@@ -0,0 +1,125 @@
+    closeSync(fd);
+    return buf.toString("utf8", 0, len);
+}
+
+export async function getDatasetRecord(
+    distributionId: string,
+    registry: Registry
+): Promise<Record | null> {
+    try {
+        const result = await registry.getRecords<Record>(
+            ["dataset-distributions"],
+            undefined,
+            undefined,
+            true,
+            undefined,
+            ["dataset-distributions.distributions:<|" + distributionId]
+        );
+
+        if ("records" in result) {
+            return (result.records[0]) || null;
+        }
+        return null;
+    } catch (e) {
+        console.error(`Unexpected error when getting dataset record`, e);
+        return null;
+    }
+}
+
+export function formatTemporal(tc?: { start?: string; end?: string }): string {
+    if (!tc) return "";
+
+    const toDate = (s?: string) => (s ? new Date(s.replace(" ", "T")) : undefined);
+    const fmt = (d: Date) =>
+        d.toLocaleDateString("en-AU", { day: "numeric", month: "short", year: "numeric" });
+
+    const start = toDate(tc.start);
+    const end = toDate(tc.end);
+
+    if (start && end) {
+        return `from ${fmt(start)} to ${fmt(end)}`;
+    }
+
+    if (start) return `from ${fmt(start)}`;
+    if (end) return `up to ${fmt(end)}`;
+    return "";
+}
+
+export const createEmbeddingText: CreateEmbeddingText = async ({
+    record,
+    format,
+    filePath,
+    url,
+    readonlyRegistry
+}) => {
+    const dist = record.aspects?.["dcat-distribution-strings"] ?? {};
+    let title = (dist.title ?? "").trim();
+    let description = (dist.description ?? "").trim();
+    const licence = (dist.license ?? "").trim();
+    const fileName = url ? basename(url.split("?")[0]) : "";
+    let temporalCoverage = "";
+    let publisher = "";
+    let themes = [];
+    let keywords = [];
+    let languages = [];
+
+    if (readonlyRegistry) {
+        const dataset = await getDatasetRecord(
+            record.id,
+            readonlyRegistry
+        );
+
+        const dsStr = dataset?.aspects?.["dcat-dataset-strings"];
+
+        if (dsStr) {
+            if (!title && dsStr.title) title = dsStr.title.trim();
+            if (!description && dsStr.description) description = dsStr.description.trim();
+            if (!publisher && dsStr.publisher) publisher = dsStr.publisher.trim();
+
+            themes = Array.isArray(dsStr.themes) ? dsStr.themes.filter(Boolean) : [];
+            keywords = Array.isArray(dsStr.keywords) ? dsStr.keywords.filter(Boolean) : [];
+            languages = Array.isArray(dsStr.languages) ? dsStr.languages.filter(Boolean) : [];
+
+            const tc = dsStr.temporal;
+            if (tc?.start || tc?.end) {
+                temporalCoverage = formatTemporal(tc);
+            }
+        }
+    }
+
+    let columns: string[] = [];
+    if (filePath && existsSync(filePath)) {
+        const header = readFirstBytes(filePath, 64 * 1024).split(/\r?\n/)[0] ?? "";
```

Comment:
For CSV indexer, probably not appropriate to let indexer framework to auto-download the full CSV and read as local file here as the size of CSV file could range from a few MBs to a few 100s MBs --- but only the columns (the first a few bytes) is what we after. If we let full CSV file download, that would be too slow and too much traffic. We probably can adopt the similar approach here: https://github.com/magda-io/magda-minion-visualization/blob/main/src/onRecordFound.ts
i.e.:
- fetch remote CSV file as a stream with [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController/abort) 
- pipe the the stream to a CSV parser that emit rows
- Once the column row is emitted or timeout, cancel the downloading via AbortController signal